wpep0002.md

```diff
@@ -0,0 +1,159 @@
+
+A PoC https://github.com/wazo-platform/xivo-lib-python/pull/37 adds a helper to
+convert configuration stored in consul key/value storage into the expected yaml
+format. It leverages the special wazo '!exec' yaml tag to load this yaml into
+the application automatically.
+
+The PoC has shown some issues, since there is no data validation and date type
+coercion, and configuration values in consul are stored as bytes, only
+configuration keys that expect a string value worked.
+Also the PoC does not support values in the form of a list.
+
+#### Ideas / Thoughts
+
+* We first need something to validate and coerce the configuration by creating
+  schemas of what we expect (with voluptuous/marshmallow/pydantic/...)
+
+* Instead of reading configuration directly from consul, we can also read it
+  from environment variables and use `envconsul` to convert consul key/value in
+  environment variables.
+  This is a really good and generic solution, which makes our applications
+  friendly with any system that generates environment for us.
+  For `Systemd` based deployment we can create the EnvironmentFile=/etc/default/wazo-XXXXX with `envconsul`
+  For `docker-compose` deployment we can create `.env` file with `envconsul`
+  For `Kubenetes` deployment we can generate the `ConfigMap` with `consul-template`
+
+  It's not fully compatible with use-cases 2.
+
+  But this gets us all container orchestration restart/update features of these
+  tools for free.
+
+https://github.com/hashicorp/envconsul
+https://github.com/hashicorp/consul-template
+
+### 2. Updating the configuration dynamically at runtime.
+
+It's clear that not all configurations can be reloaded at runtime.
+
+Most of the configuration is used when the application starts (ip/port binding,
+ssl certificates, rabbit and postgres settings, enabled plugins, etc...).
+The application startup instantiates many Python classes (http/database/bus
+clients/helpers) that use this configuration settings and they are passed around the
+application, making it hard to change them at runtime.
+
+Also we need to be able to detect when something is changed. python-consul
+provides watch ability on key/value store.
+If a configuration option is always read from the startup-built config dict,
+the application will automatically use the new value, but if a value was
+duplicated inside a class somewhere in the app, then we need specific code to
+propagate the config change to this class.
+
+#### Ideas / Thought
+
+* Most of the configuration options are not worth the cost of writing specific code, like
+  changing ip/port of postgres/rabbit/http layers that would interrupt the
+  services anyway as an application/containers restart would.
+
+* Don't support dynamic configuration change within the application, and just
+  restart the application/container when the configuration is updated. This
+  makes the applications more kubernetes/docker-compose friendly, especially if
```

Comment:
Speaking of Kubernetes/docker-compose, being able to specify the configuration using environment variables would be very useful. For instance, if the configuration file is:
```
---
section1:
   key1: myvalue
```
setting the environment variable WAZO_SECTION1_KEY=mynewvalue would override the previous `myvalue`